rfc-128-continuous-deployment.md

```diff
@@ -0,0 +1,223 @@
+
+- We only need to add contract tests for the APIs we consume internally. Some of our APIs may be public, but this is a separate, product-level commitment. Our strategy for testing public APIs is out-of-scope for this RFC.
+
+With this in mind, an API app has "enough" of these tests when:
+
+- Each endpoint with multiple consumers has at least one contract test.
+
+> Example: [Asset Manager](https://github.com/alphagov/asset-manager) is missing contract tests because multiple apps use it to create [[1](https://github.com/alphagov/content-publisher/blob/0c757447ca2aad3621f11c99fe8307a718ade186/app/services/preview_asset_service.rb#L29)] [[2](https://github.com/alphagov/specialist-publisher/blob/8debeb2f0147142c87f22308d57fe4f6dd0c1297/app/models/attachment.rb#L49)] assets, among other APIs. Conversely, no [Support API](https://github.com/alphagov/support-api) endpoints are used by more than one GOV.UK app [[1](https://github.com/alphagov/support-api/tree/75540575ae11095bb8d4736490abdc9927049aaf#technical-documentation)].
+
+#### Check: Smoke tests pass (i.e. the app runs in a production environment)
+
+For the purpose of automating deployments:
+
+- We need to test that an app can respond to a request for each of the machine classes it runs on [[1](https://github.com/alphagov/govuk-app-deployment/blob/d4bb62aa79face46d2e164e316f38f7a64a1b8be/content-store/config/deploy.rb#L3)] [[2](https://github.com/alphagov/govuk-app-deployment/blob/d4bb62aa79face46d2e164e316f38f7a64a1b8be/whitehall/config/deploy.rb#L13)]. This is purely to prove that an instance of the app is running [[1](https://github.com/alphagov/email-alert-api/pull/1346)], and may be covered by other tests.
+
+- We need to test that an app can connect to any systems it uses for data storage and retrieval, such as a database. This usually has production-specific config [[1](https://github.com/alphagov/email-alert-api/blob/1b5f59f3927637afb0c790702f8f0050788edc9d/config/database.yml#L22)] [[2](https://github.com/alphagov/content-store/blob/f4b61d4f2756341b6418085d878ec83b09d70e9c/config/mongoid.yml#L23)] [[3](https://github.com/alphagov/content-publisher/blob/944fe51b4cb8205348e155b965b8bad57adcf4fe/config/storage.yml#L5)], which could be faulty.
+
+- We do not need to test connectivity to other GOV.UK apps, since this should be handled by [our internal API adapters](https://github.com/alphagov/gds-api-adapters/blob/1ceba690edb37295257328d953874224833fcd5e/lib/gds_api.rb) with no configuration inside the repo. An exception to this is [authentication with Signon](https://github.com/alphagov/email-alert-api/blob/1b5f59f3927637afb0c790702f8f0050788edc9d/config/initializers/gds_sso.rb).
+
+With this in mind, an app has "enough" of these tests when:
+
+- It has a test that checks for a successful response to an arbitrary request.
+- It has a test that checks for a successful login with Signon (if applicable).
```

Comment:
Thanks for the chat about this. I originally intended this to be another Smokey test, but you rightly pointed out just how many extra environment variable this would require us to bestow on that repo - one for each token.

The healthcheck suggestion is interesting. While I agree an app is very broken if its SSO config is broken, doing that check as part of the `/healthcheck` endpoint is oddly separate to the other proposed here.

Thinking about it some more, I actually think we can avoid all Signon checks, by simply removing the code we're concerned about from the repos. I've got a PR to do that [here](https://github.com/alphagov/gds-sso/pull/241), could you take a look?

Reply:
Yup, that sounds a good idea. I've dropped a review there. Nice one.

Reply:
Thanks! We've started rolling out this change, so I think it's safe to remove it as a criterion: https://github.com/alphagov/govuk-rfcs/pull/128/commits/0784737b001085dd5d0ec3536cc88bf37adc57b3.